react/states_and_effects/more_on_state.md

```diff
@@ -0,0 +1,213 @@
+      setPerson(newPerson);
+   }
+   
+   return (
+      <>
+         <h1>{person.name}</h1>
+         <h2>{person.age}</h2>
+         <button onClick={handleIncreaseAge}>Increase age</button>
+      </>
+   )
+}
+~~~
+
+<div class="lesson-note lesson-note--tip" markdown="1">
+
+#### Nested objects and arrays in state
+
+In the above example, notice how we had to copy the existing state object. Herein lies another lesson, nested objects and arrays as state can get tricky fast, since you will have to replicate the nested items as well. Be careful when using them.
+
+</div>
+
+### How state updates
+
+State updates are asynchronous. What this implies is whenever you call the `setState` function, React will apply the update in the **next** component render. This concepts takes a while to wrap your head around. With a lot of practice, you'll get the hand of it no time.
+
+Remember, state variables aren't reactive, the component is. This can be understood by the fact that calling `setState` re-renders the entire component instead of just changing the state variable on the fly.
+
+~~~jsx
+function Person() {
+   const [person, setPerson] = useState({ name: 'John', age: 100 });
+
+   const handleIncreaseAge = () =>{
+      console.log("in handleIncreaseAge (before setPerson call): ", person)
+      setPerson({ ...person, age: person.age + 1 });
+      // we've called setPerson, surely person has updated?
+      console.log("in handleIncreaseAge (after setPerson call): ", person);
+   }
+
+   // this console.log runs every time the component renders 
+   // what do you think this will print?
+   console.log("during render: ", person);
+   
+   return (
+      <>
+         <h1>{person.name}</h1>
+         <h2>{person.age}</h2>
+         <button onClick={handleIncreaseAge}>Increase age</button>
+      </>
+   )
+} 
+~~~
+
+These are the logs:
+
+![](./more_on_state/imgs/00.png)
+
+Uh oh, what is happening? Let's break it down (ignore the double `console.logs` for the render case, this is covered in the upcoming lessons).
+
+1. The component renders for the first time. The `person` state variable is initialized to `{ name: 'John', age: 100 }`. The "during render" `console.log` prints the state variable.
+1. The button is clicked invoking `handleIncreaseAge`. Interestingly, the `console.log` before and after the `setPerson` call print the same value.
+1. The component re-renders. The `person` state variable is updated to `{ name: 'John', age: 101 }`. 
+
+The `person` state stays the same throughout the current render of the component. This is what "state as a snapshot" refers to. The `setState` call triggers a component re-render and the `person` state is updated to the new value.
+
+<div class="lesson-note lesson-note--warning" markdown="1">
+
+#### The unexpected infinite loop
+
+The following is an infinite loop, can you guess why? Drop by in our [Discord chatroom](https://discord.com/invite/fbFCkYabZB), tell us why, and score a brownie point!
```

Comment:
lol will there be a brownie bot update command 😂?

Reply:
we can add one 🤣